### DISCUSSION TITLE
Let Transforms and Linters operate on rendered Node Buffers

### DISCUSSION THREAD

**Discussion by jordanthornquest** [11ty/eleventy#4080] [Enhancement Queue]
**Describe the bug**
It appears that transforms don't quite work as I expected. I'm using JavaScript templates for rendering my site's pages, and returning Buffers for rendering (per stated compatibility on [Eleventy's docs](https://www.11ty.dev/docs/languages/javascript/#classes)). I'm using [`html-minifier`](https://github.com/kangax/html-minifier) to minify the HTML output.  However, when returning a Buffer, `html-minifier` returns a TypeError, saying that `str.replace` is not a function.

Presumably, this is because the transform is being run on a Buffer object rather than a String. I had assumed that transforms were done on the HTML output generated by Eleventy, rather than the objects passed from templates. Is this incorrect?

**To Reproduce**

1. Clone [this website repository](https://github.com/jordanthornquest/jordanthornquest/tree/9c90e5e011bb4b4eea5ce1077462d26b98621c95) (linked to specific commit in Git history) and `npm install`.
2. `npm run build` to compile site content. This will minify site output.
    * Minification is handled by a transform, which can be found at `src/_utils/transforms/htmlmin.js`
3. Open `src/_layouts/base.11ty.js`
4. Change the following lines:
    ```js
    // Change this line (line 2)
    const { html, renderToString } = require("@popeindustries/lit-html-server");
    // To
    const { html, renderToBuffer } = require("@popeindustries/lit-html-server");

    // Change this line (line 131)
    const renderedLayout = await renderToString(layoutTemplate);
    // To
    const renderedLayout = await renderToBuffer(layoutTemplate);
    ```
5. `npm run build` to compile site content. `html-minifier` will return a TypeError.

**Expected behavior**
I would expect the behavior to be identical when using either function, since I had assumed that transforms were done to the HTML output.

**Environment:**

- MacOS 11.3.1 (M1 MacBook Air)
- Eleventy `v0.12.1`
- Node `v16.0.0`

**Additional context**
I'm choosing to return Buffers instead of Strings because doing so circumvents the extra step of escaping HTML within strings when rendering templates and layouts. I'm rendering HTML into Buffers using `[lit-html-server](https://github.com/popeindustries/lit-html-server)`.

**Comment by zachleat** [comment]
A ha, I see what you mean.

Eleventy does convert the rendered template to string before it passes to linters and transforms. Is that a bug? Is this a feature request? I’m not sure! I am inclined to say feature request.

I will say that it is more convenient to do this as it simplifies the transform code.

https://github.com/11ty/eleventy/blob/d19257350887c9f670f8652039555d9460f69f79/src/Template.js#L755

https://github.com/11ty/eleventy/blob/d19257350887c9f670f8652039555d9460f69f79/src/Engines/JavaScript.js#L16

If we do change this, I’m guessing it might cause some confusion from folks that were relying on it so it may need to be opt-in.

**Comment by zachleat** [comment]
> This repository is now using [lodash style issue management](https://twitter.com/samselikoff/status/991395669016436736) for enhancements. This means enhancement issues will now be closed instead of leaving them open.

> View [the enhancement backlog here](https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+). Don’t forget to upvote the top comment with 👍!
